extensions/kafka-eight-simpleConsumer/src/main/java/io/druid/firehose/kafka/KafkaEightSimpleConsumerFirehoseFactory.java

```diff
@@ -0,0 +1,318 @@
+      log.error("Unable to cast lastCommit to Map");
+    }
+    return offsetMap;
+  }
+
+  @Override
+  public FirehoseV2 connect(final ByteBufferInputRowParser firehoseParser, Object lastCommit) throws IOException
+  {
+    final Map<Integer, Long> lastOffsets = loadOffsetFromPreviousMetaData(lastCommit);
+
+    Set<String> newDimExclus = Sets.union(
+        firehoseParser.getParseSpec().getDimensionsSpec().getDimensionExclusions(),
+        Sets.newHashSet("feed")
+    );
+    final ByteBufferInputRowParser theParser = firehoseParser.withParseSpec(
+        firehoseParser.getParseSpec()
+                      .withDimensionsSpec(
+                          firehoseParser.getParseSpec()
+                                        .getDimensionsSpec()
+                                        .withDimensionExclusions(
+                                            newDimExclus
+                                        )
+                      )
+    );
+    for (Integer partition : partitionIdList) {
+      final KafkaSimpleConsumer kafkaSimpleConsumer = new KafkaSimpleConsumer(
+          feed, partition, clientId, brokerList, earliest
+      );
+      Long startOffset = lastOffsets.get(partition);
+      PartitionConsumerWorker worker = new PartitionConsumerWorker(
+          feed, kafkaSimpleConsumer, partition, startOffset == null ? 0 : startOffset
+      );
+      consumerWorkers.add(worker);
+    }
+
+    final LinkedBlockingQueue<BytesMessageWithOffset> messageQueue = new LinkedBlockingQueue<BytesMessageWithOffset>(queueBufferLength);
+    log.info("Kicking off all consumers");
+    for (PartitionConsumerWorker worker : consumerWorkers) {
+      worker.go(messageQueue);
+    }
+    log.info("All consumer started");
+
+    return new FirehoseV2()
+    {
+      private ConcurrentMap<Integer, Long> lastOffsetPartitions;
+      private volatile boolean stop;
+      private volatile boolean interrupted;
+
+      private volatile BytesMessageWithOffset msg = null;
+      private volatile InputRow row = null;
+
+      {
+        lastOffsetPartitions = Maps.newConcurrentMap();
+        lastOffsetPartitions.putAll(lastOffsets);
+      }
+
+      @Override
+      public void start() throws Exception
+      {
+        nextMessage();
+      }
+
+      @Override
+      public boolean advance()
+      {
+        if (stop) {
+          return false;
+        }
+
+        nextMessage();
+        return true;
+      }
+
+      private void nextMessage()
+      {
+        try {
+          row = null;
+          while (row == null) {
+            if (msg != null) {
+              lastOffsetPartitions.put(msg.getPartition(), msg.offset());
+            }
+
+            msg = messageQueue.take();
+            interrupted = false;
+
+            final byte[] message = msg.message();
+            row = message == null ? null : theParser.parse(ByteBuffer.wrap(message));
+          }
+        }
+        catch (InterruptedException e) {
+          interrupted = true;
```

Comment:
`InterruptedExcpetion` is really a temporary state, and we hope to recover message consumption in future `nextMessage()`, and in case of firehose shutdown, we will pass the signal through `advance()`, so we actually do not need higher level methods to be aware of that.